decred/decred/dcr/account.py

```diff
@@ -833,19 +1050,173 @@ def calcBalance(self, tipHeight=None):
+        if tinfo.vote or tinfo.revocation:
+            return tinfo
+        log.debug("ticket's tinfo not found on the blockchain {}".format(txid))
+
+    def spendTicket(self, tx):
+        """
+        Update a ticket that is already in the ticketDB. Sends a
+            ui.SPENT_TICKETS_SIGNAL on completion.
+
+        Args:
+            tx (msgtx.MsgTx): the vote or transaction that spends a ticket in
+                the ticketDB.
+        """
+        isVote = txscript.isSSGen(tx)
+        idx = 1 if isVote else 0
+        ticketTxid = tx.txIn[idx].previousOutPoint.txid()
+        if ticketTxid not in self.ticketDB:
+            raise Exception(
+                "spending tx came for ticket that was not found in the database: {}".format(
+                    ticketTxid
+                )
+            )
+        tinfo = self.blockchain.ticketInfoForSpendingTx(tx.txid(), self.net)
+
+        ticket = self.ticketDB[ticketTxid]
+        ticket.tinfo = tinfo
+        self.ticketDB[ticketTxid] = ticket
+
+    def updateSpentTickets(self, poolTxids=None):
+        """
+        Update voted and revoked tickets. ui.SPENT_TICKETS_SIGNAL is called if
+        there was an update.
+
+            Args:
+                poolTxids (list(str)): Optional. Default=None. Additional txids
+                    to check for tickets.
+        """
+        # Get current unspent tickets.
+        mempoolTickets = {
+            k: UTXO.ticketFromTx(v, self.net)
+            for k, v in self.mempool.items()
+            if v.isTicket()
+        }
+        activeTickets = {
+            utxo.txid: utxo for utxo in self.utxos.values() if utxo.isTicket()
+        }
+        activeTickets.update(mempoolTickets)
+        # Get tickets as stored in the database.
+        unspentTickets, voted, revoked, unconfs = self.sortedTickets()
+
+        # Check tickets with unconfirmed status.
+        checkers = unconfs
+
+        for txid, tkt in activeTickets.items():
+            # Check new tickets.
+            if txid not in unspentTickets:
+                checkers.append(txid)
+                continue
+            dbTicket = unspentTickets[txid]
+            # Check tickets whose status has changed.
+            if dbTicket.tinfo.status != tkt.tinfo.status:
+                checkers.append(txid)
+
+        for txid in unspentTickets.keys():
+            # Check tickets that have been spent or disappeared due to a reorg
+            # or never have being mined.
+            if txid not in activeTickets:
+                checkers.append(txid)
+
+        # Check for possible unknown tickets if pool txids are supplied.
+        if poolTxids:
+            for txid in poolTxids:
+                if txid not in voted and txid not in revoked:
+                    checkers.append(txid)
+
+        remove = []
+        for txid in checkers:
+            try:
+                tx = self.blockchain.tx(txid)
+            except Exception:
```

Comment:
The TODO already there. Moved it a little tho. This should definitely be done because the ticket db will will be wiped if the blockchain is disconnected here.